base/compiler/ssair/inlining2.jl

```diff
@@ -484,123 +816,99 @@ function assemble_inline_todo!(ir::IRCode, linetable::Vector{LineInfoNode}, sv::
-        # Bail out if any static parameters are left as TypeVar
-        ok = true
-        for i = 1:length(methsp)
-            isa(methsp[i], TypeVar) && (ok = false; break)
-        end
-        ok || continue
-
-        # Find the linfo for this methods (Generated functions are expanded here if necessary)
-        linfo = code_for_method(method, metharg, methsp, sv.params.world, true) # Union{Nothing, MethodInstance}
-        if !isa(linfo, MethodInstance)
-            maybe_make_invoke!(ir, idx, stmt.typ, atypes, sv, atype_unlimited, isinvoke, isapply, invoke_data)
-            continue
-        end
+        cases = Pair{Type, Any}[]
+        # TODO: This could be better
+        signature_union = Union{Any[match[1]::Type for match in meth]...}
+        signature_fully_covered = atype <: signature_union
+        fully_covered = signature_fully_covered
+        split_out_sigs = Any[]
 
-        if invoke_api(linfo) == 2
-            # in this case function can be inlined to a constant
-            add_backedge!(linfo, sv)
-            ir[SSAValue(idx)] = linfo.inferred_const
-            continue
+        # For any method match that's a dispatch tuple, extract those cases first
+        for (i, match) in enumerate(meth)
+            (metharg, methsp, method) = (match[1]::Type, match[2]::SimpleVector, match[3]::Method)
+            if !isdispatchtuple(metharg)
+                fully_covered = false
+                continue
+            end
+            case = analyze_method!(idx, f, ft, metharg, methsp, method, stmt, atypes, sv, metharg, isinvoke, isapply, invoke_data)
+            if case === nothing
+                fully_covered = false
+                continue
+            end
+            push!(cases, Pair{Type,Any}(metharg, case))
+            push!(split_out_sigs, metharg)
         end
 
-        # Handle vararg functions
-        isva = na > 0 && method.isva
-        if isva
-            @assert length(atypes) >= na - 1
-            va_type = tuple_tfunc(Tuple{Any[widenconst(atypes[i]) for i in 1:length(atypes)]...})
-            atypes = Any[atypes[1:(na - 1)]..., va_type]
+        # Now, if profitable union split the atypes into dispatch tuples and match the appropriate method
+        nu = countunionsplit(atypes)
+        if nu != 1 && nu <= sv.params.MAX_UNION_SPLITTING
+            for sig in UnionSplitSignature(atypes)
+                metharg′ = argtypes_to_type(sig)
+                if !isdispatchtuple(metharg′)
+                    fully_covered = false
+                    continue
+                elseif any(x->x === metharg′, split_out_sigs)
+                    continue
+                end
+                # `meth` is in specificity order, so find the first applicable method
+                found_any = false
+                for (i, match) in enumerate(meth)
+                    (metharg, methsp, method) = (match[1]::Type, match[2]::SimpleVector, match[3]::Method)
+                    metharg′ <: metharg || continue
```

Comment:
type-intersect

Reply:
Isn't that equivalent, since `metharg′` is a dispatch tuple?

Reply:
ah yes, although that `metharg` is not relevant to whether the method will actually match at runtime

Reply:
Could you elaborate that statement?

Reply:
metharg is `typeintersect(method.sig, atypes)`, so it's unpredictable whether it's `method.sig`, `atypes`, or something wider or narrower. basically, just assume that it's `Tuple` and you should be OK.

Reply:
Ok, so should that be method.sig on the RHS of this check then?

Reply:
yeah, that should work